// react-native-ui-lib/src/components/sortableGridList/index.js
import _map from "lodash/map";
import _pt from "prop-types";
import React, { useCallback } from 'react';
import { StyleSheet, ScrollView } from 'react-native';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import { useSharedValue } from 'react-native-reanimated';
import SortableItem from "./SortableItem";
import usePresenter from "./usePresenter";
import useGridLayout, { DEFAULT_ITEM_SPACINGS, DEFAULT_NUM_COLUMNS } from "../gridList/useGridLayout";

function SortableGridList(props) {
  const {
    renderItem,
    onOrderChange,
    contentContainerStyle,
    ...others
  } = props;
  const {
    itemContainerStyle,
    numberOfColumns
  } = useGridLayout(props);
  const {
    numColumns = DEFAULT_NUM_COLUMNS,
    itemSpacing = DEFAULT_ITEM_SPACINGS,
    data
  } = others;
  const itemsOrder = useSharedValue(_map(props.data, (_v, i) => i)); // TODO: Get the number of columns from GridList calculation somehow

  const presenter = usePresenter(props.data?.length ?? 0, numColumns, itemSpacing);
  const onChange = useCallback(() => {
    const newData = [];

    if (data?.length) {
      itemsOrder.value.forEach(itemIndex => {
        newData.push(data[itemIndex]);
      });
    }

    onOrderChange?.(newData, itemsOrder.value);
  }, [onOrderChange, data]);

  const _renderItem = useCallback(({
    item,
    index
  }) => {
    const lastItemInRow = (index + 1) % numberOfColumns === 0;
    return <SortableItem key={index} {...presenter} style={[itemContainerStyle, lastItemInRow && {
      marginRight: 0
    }]} itemsOrder={itemsOrder} index={index} onChange={onChange}>
        {
        /* @ts-expect-error */
      }
        {renderItem({
        item,
        index
      })}
      </SortableItem>;
  }, []);

  return <GestureHandlerRootView>
      <ScrollView contentContainerStyle={[styles.listContent, contentContainerStyle]}>
        {data?.map((item, index) => _renderItem({
        item,
        index
      }))}
      </ScrollView>
    </GestureHandlerRootView>;
}

SortableGridList.propTypes = {
  onOrderChange: _pt.func
};
export default SortableGridList;
const styles = StyleSheet.create({
  listContent: {
    flexWrap: 'wrap',
    flexDirection: 'row'
  }
});
