// react-native-ui-lib/src/components/sortableGridList/usePresenter.js
import _times from "lodash/times";
import { Easing, useSharedValue } from 'react-native-reanimated';
import Constants from "../../commons/Constants";
export const WINDOW_WIDTH = Constants.windowWidth;
export const DEFAULT_NO_OF_COLUMNS = 3;
export const getItemSize = (numOfColumns, viewWidth) => viewWidth / numOfColumns;
export const animationConfig = {
  easing: Easing.inOut(Easing.ease),
  duration: 350
};

const usePresenter = (itemsCount, numOfColumns, itemSpacing) => {
  const itemsLayouts = useSharedValue(_times(itemsCount, () => undefined));
  const itemSize = getItemSize(numOfColumns, Constants.screenWidth);
  return {
    updateItemLayout: (index, layout) => {
      'worklet';

      itemsLayouts.value[index] = layout;

      if (itemsLayouts.value.findIndex(item => item === undefined) === -1) {
        itemsLayouts.value = [...itemsLayouts.value];
      }
    },
    getTranslationByOrderChange: (newOrder, oldOrder) => {
      'worklet';

      const oldRow = Math.floor(oldOrder / numOfColumns);
      const oldColumn = oldOrder % numOfColumns;
      const newRow = Math.floor(newOrder / numOfColumns);
      const newColumn = newOrder % numOfColumns;
      const translation = {
        x: (newColumn - oldColumn) * ((itemsLayouts.value[0]?.width ?? 0) + itemSpacing),
        y: (newRow - oldRow) * ((itemsLayouts.value[0]?.height ?? 0) + itemSpacing)
      };
      return translation;
    },
    getOrderByPosition: (positionX, positionY) => {
      'worklet';

      const col = Math.round(positionX / itemSize);
      const row = Math.round(positionY / itemSize);
      return row * numOfColumns + col;
    },
    getItemOrderById: (itemsOrder, itemId) => {
      'worklet';

      return itemsOrder.indexOf(itemId);
    },
    getIdByItemOrder: (itemsOrder, orderIndex) => {
      'worklet';

      return itemsOrder[orderIndex];
    }
  };
};

export default usePresenter;
