// react-native-ui-lib/src/components/button/Button.driver.js
import _get from "lodash/get";
import { fireEvent } from '@testing-library/react-native';
import { TextDriver } from "../text/Text.driver";
import { ImageDriver } from "../image/Image.driver";
export const ButtonDriver = async ({
  wrapperComponent,
  testID
}) => {
  const button = await wrapperComponent.queryByTestId(testID);
  const label = await TextDriver({
    wrapperComponent,
    testID: `${testID}.label`
  });
  const icon = await ImageDriver({
    wrapperComponent,
    testID: `${testID}.icon`
  });
  return {
    exists: () => !!button,
    getRootElement: () => button,
    isClickable: () => !_get(button, 'props.accessibilityState.disabled'),
    click: () => {
      if (button) {
        fireEvent.press(button);
      } else {
        console.warn(`ButtonDriver: cannot click because testID:${testID} were not found`);
      }
    },
    // label
    getLabelRootElement: () => label.getRootElement(),
    isLabelExists: () => label.exists(),
    getLabelContent: () => label.getTextContent(),
    // icon
    getIconRootElement: () => icon.getRootElement(),
    isIconExists: () => icon.exists()
  };
};
