// react-native-ui-lib/src/incubator/TextField/TextField.driver.js
import _isEmpty from "lodash/isEmpty";
import _get from "lodash/get";
import { fireEvent } from '@testing-library/react-native';
import { TextDriver } from "../../components/text/Text.driver";
export const TextFieldDriver = async ({
  wrapperComponent,
  testID
}) => {
  const textInput = await wrapperComponent.queryByTestId(testID);
  const label = await TextDriver({
    wrapperComponent,
    testID: `${testID}.label`
  });
  const validationMsg = await TextDriver({
    wrapperComponent,
    testID: `${testID}.validationMessage`
  });
  const floatingPlaceholder = await TextDriver({
    wrapperComponent,
    testID: `${testID}.floatingPlaceholder`
  });
  const charCounter = await TextDriver({
    wrapperComponent,
    testID: `${testID}.charCounter`
  });

  function isPlaceholderVisible() {
    if (textInput) {
      const hasPlaceholderProp = !!_get(textInput, 'props.placeholder');
      const hasInputText = !!_get(textInput, 'props.value');
      return hasPlaceholderProp && (!hasInputText || hasInputText && floatingPlaceholder.exists());
    } else {
      console.warn(`TextField component with testId:${testID}, is not found. So you can't get his placeholder`);
    }
  }

  return {
    exists: () => !!textInput,
    getRootElement: () => textInput,
    content: () => {
      if (textInput) {
        return textInput.props.value;
      } else {
        console.warn(`TextField component with testId:${testID}, is not found. So you can't get the content`);
        return null;
      }
    },
    isDisabled: () => {
      if (textInput) {
        return _get(textInput, 'props.accessibilityState.disabled');
      } else {
        console.warn(`TextField component with testId:${testID}, is not found. So you can't get the content`);
        return null;
      }
    },
    changeText: text => {
      if (textInput) {
        fireEvent.changeText(textInput, text);
      } else {
        console.warn(`TextFieldDriver: cannot change text because testID:${testID} were not found`);
      }
    },
    // placeholder
    isPlaceholderVisible,
    getPlaceholderContent: () => {
      if (isPlaceholderVisible()) {
        return _get(textInput, 'props.placeholder');
      } else {
        console.warn(`You cant get placeholder content, cause placeholder is not visible.`);
        return null;
      }
    },
    // label
    getLabelRootElement: () => label.getRootElement(),
    isLabelExists: () => label.exists() && !floatingPlaceholder.exists(),
    getLabelContent: () => label.getTextContent(),
    // validation message
    getValidationMsgRootElement: () => validationMsg.getRootElement(),
    isValidationMsgExists: () => validationMsg.exists() && !_isEmpty(validationMsg.getTextContent()),
    getValidationMsgContent: () => validationMsg.getTextContent(),
    //leadingAccessory, trailingAccessory, bottomAccessory
    // char counter
    getCharCounterRootElement: () => charCounter.getRootElement(),
    isCharCounterExists: () => charCounter.exists(),
    getCharCounterContent: () => charCounter.getTextContent()
  };
};
